relay/writer/file.go

```diff
@@ -0,0 +1,381 @@
+		// if not open a binlog file yet, then non-fake RotateEvent can't be handled
+		return nil, errors.Errorf("non-fake RotateEvent %+v received, but no binlog file opened", ev.Header)
+	} else {
+		err := w.out.Write(ev.RawData)
+		if err != nil {
+			return nil, errors.Annotatef(err, "write RotateEvent %+v for %s", ev.Header, filepath.Join(w.cfg.RelayDir, currFile))
+		}
+	}
+
+	return &Result{
+		Ignore: ignore,
+	}, nil
+}
+
+// handle non-special event:
+//   1. handle a potential hole if exists
+//   2. handle any duplicate events if exist
+//   3. write the non-duplicate event
+func (w *FileWriter) handleEventDefault(ev *replication.BinlogEvent) (*Result, error) {
+	// handle a potential hole
+	err := w.handleFileHoleExist(ev)
+	if err != nil {
+		return nil, errors.Annotatef(err, "handle a potential hole in %s before %+v",
+			w.filename.Get(), ev.Header)
+	}
+
+	// handle any duplicate events if exist
+	result, err := w.handleDuplicateEventsExist(ev)
+	if err != nil {
+		return nil, errors.Annotatef(err, "handle a potential duplicate event %+v in %s",
+			ev.Header, w.filename.Get())
+	}
+	if result.Ignore {
+		// duplicate, and can ignore it. now, we assume duplicate events can all be ignored
+		return result, nil
+	}
+
+	// write the non-duplicate event
+	err = w.out.Write(ev.RawData)
+	return &Result{
+		Ignore: false,
+	}, errors.Annotatef(err, "write event %+v", ev.Header)
+}
+
+// handleFileHoleExist tries to handle a potential hole after this event wrote.
+// A hole exists often because some binlog events not sent by the master.
+// NOTE: handle cases when file size > 4GB
+func (w *FileWriter) handleFileHoleExist(ev *replication.BinlogEvent) error {
+	// 1. detect whether a hole exists
+	evStartPos := int64(ev.Header.LogPos - ev.Header.EventSize)
+	outFs, ok := w.out.Status().(*bw.FileWriterStatus)
+	if !ok {
+		return errors.Errorf("invalid status type %T of the underlying writer", w.out.Status())
+	}
+	fileOffset := outFs.Offset
+	holeSize := evStartPos - fileOffset
+	if holeSize <= 0 {
+		// no hole exists, but duplicate events may exists, this should be handled in another place.
+		return nil
+	}
+
+	// 2. generate dummy event
+	var (
+		header = &replication.EventHeader{
+			Timestamp: uint32(time.Now().Unix()),
+			ServerID:  ev.Header.ServerID,
+		}
+		latestPos = uint32(fileOffset)
+		eventSize = uint32(holeSize)
+	)
+	dummyEv, err := event.GenDummyEvent(header, latestPos, eventSize)
+	if err != nil {
+		return errors.Annotatef(err, "generate dummy event at %d with size %d", latestPos, eventSize)
+	}
+
+	// 3. write the dummy event
+	err = w.out.Write(dummyEv.RawData)
+	return errors.Trace(err)
+}
+
+// handleDuplicateEventsExist tries to handle a potential duplicate event in the binlog file.
+func (w *FileWriter) handleDuplicateEventsExist(ev *replication.BinlogEvent) (*Result, error) {
+	filename := filepath.Join(w.cfg.RelayDir, w.filename.Get())
+	duplicate, err := checkIsDuplicateEvent(filename, ev)
```

Comment:
is it inefficient that call it for every event?

Reply:
in most cases, `checkIsDuplicateEvent` only checks start/end pos with the file size (has a comment in `checkIsDuplicateEvent`), so I think it is not inefficient. and if we want to make it work correctly, we must do this check.

Reply:
I'm worrid that `os.Stat` is an inefficient function

Reply:
so, pass file size into the function as an arg?

Reply:
for `os.Stat`, it will call `syscall.Stat`; for `File.Stat`, it will call `syscall.Fstat`.

from <http://man7.org/linux/man-pages/man2/stat.2.html>.
> fstat() is identical to stat(), except that the file about which
       information is to be retrieved is specified by the file descriptor
       fd.

if we changed to pass `File.Stat()` into the function, the difference of efficiency is small? or should we need to pass an in-memory offset into the function? @GregoryIan what's your opinion.

Reply:
I think we can choose one
* check duplicate event only if we need to check, like `holeSize < 0`
* or parse a `fd` or `stat information` as arg?

Reply:
I choose the first one.